wagtail/admin/static_src/wagtailadmin/js/page-chooser.js

```diff
@@ -39,4 +42,17 @@ function createPageChooser(id, pageTypes, openAtParentId, canChooseRoot, userPer
         openAtParentId = null;
         chooserElement.addClass('blank');
     });
+
+    if (input.val()) {
+        $.ajax({
+            url: window.wagtailConfig.ADMIN_API.PAGES + encodeURIComponent(input.val()) + '/',
```

Comment:
This will cause an extra HTTP request for every chooser widget rendered, whether it's part of a StreamField or not (and if it isn't, the HTTP request is redundant). Chooser widgets are intended to be a standalone reusable component, so if they need a StreamField-specific customisation, then that's something we need to consider carefully.

What exactly is the technical limitation of react-streamfield that makes this change necessary? (Or, to put it another way: as a Django developer, what do I need to know to ensure that the form widgets I create will be compatible with react-streamfield?)

Is it that it has to be rendered entirely client-side, so we don't have access to any data that isn't part of the StreamField JSON (such as page title)? If so, isn't that what `prepare_value` / `prepare_for_react` are for? (How does `RichTextBlock` handle it, given that it relies heavily on server-side data conversion?)

Reply:
Yes, I’m aware that it triggers a bunch of AJAX requests on page load. Here’s why:

It is not due to a technical limitation, it’s for the new duplication feature, as well as for optimization.

When you duplicate a block, you cannot copy its full JavaScript state, you have to call that page/image/document/snippet callback again. So doing that AJAX call was an easy fix to make duplication work without rewriting the whole chooser logics. And after doing that, I was amazed to see that it was actually a lot faster than I thought. Because like you, I’m not a fan of doing a ton of async requests.

So the next step for me was to try to avoid doing the Django template rendering of the same blocks over and over again, because we could just fetch only the data we need using AJAX. And the results was even more amazing, that’s why this StreamField is considerably faster to load than the old StreamField, even when you take into account the AJAX requests. For info, there is one case where I always render a block anyway: when there is an error in the block. Otherwise I only do a single render per block type.

So after these considerations I thought: why not do it this way? Since our goal with Wagtail is to go towards a headless CMS & more dynamic page editor (like the realtime preview), this approach of AJAX widget is not heretic. And since for some extreme cases it’s up to **1000 × faster**, I became actually proud of this approach ^^

Reply:
> When you duplicate a block, you cannot copy its full JavaScript state, you have to call that page/image/document/snippet callback again. So doing that AJAX call was an easy fix to make duplication work without rewriting the whole chooser logics.

OK, so my question still stands: as a Django developer, what do I need to know to ensure that the form widgets I create will be compatible with react-streamfield's implementation of block duplication?

Reply:
The only thing to do is to have the same HTML code in the widget, regardless of the value. The only scenario where that can change is for displaying errors.